---
title: "Raw vs Normalized vs Displacement Scores"
subtitle: "Microbiome Results"
author: "Michael Sieler"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float: true
    cache: true
    cache.lazy: true
    cache.comments: false
    cache.rebuild: false
---

```{r setup, include=FALSE, cache=TRUE}
knitr::opts_chunk$set(
  echo = TRUE, 
  warning = FALSE, 
  message = FALSE,
  cache = TRUE,           # Enable caching for all chunks
  cache.lazy = TRUE,      # Lazy loading of cached objects
  cache.comments = FALSE, # Don't cache comments (saves space)
  cache.rebuild = FALSE   # Don't rebuild cache unless needed
)

# Load the here package for project-relative paths
library(here)

# Get the project root directory (where the .Rproj file is located)
proj.path <- here::here()

# Example paths to other project directories using here package
# These are more robust than using getwd() + paste0()
# path.results <- here::here("Results")
# path.data <- here::here("Data")
# path.code <- here::here("Code")
```


## Overview

```{r echo=FALSE}
# Use here package to build project-relative path to the experimental design image
knitr::include_graphics(here::here("Manuscript", "Figures", "Experimental_Design", "Experimental_Design.png"))
```


## Setup  {.tabset}

### (hide)

Click on tabs to display additional information.

```{r}

```

### Libraries

```{r message=FALSE, warning=FALSE}
# Load required libraries for statistical analysis and table creation
library(dplyr)
library(ggplot2)
library(gt)
library(broom)
library(car)
library(emmeans)
library(multcomp)
library(MASS)  # For negative binomial regression
library(rstatix)
library(coin)
library(phyloseq)
library(microViz)
library(tidyverse)
```

### Plotting

```{r}

# Define treatment order and color palette
treatment_order <- c(
  "A- T- P-",  # Control
  "A- T- P+",  # Parasite
  "A+ T- P-",  # Antibiotics
  "A+ T- P+",  # Antibiotics_Parasite
  "A- T+ P-",  # Temperature
  "A- T+ P+",  # Temperature_Parasite
  "A+ T+ P-",  # Antibiotics_Temperature
  "A+ T+ P+"   # Antibiotics_Temperature_Parasite
)

# Custom color palette matching treatment order
treatment_colors <- c(
  "#1B9E77",  # A- T- P- (Control)
  "#D95F02",  # A- T- P+ (Parasite)
  "#7570B3",  # A+ T- P- (Antibiotics)
  "#E7298A",  # A+ T- P+ (Antibiotics_Parasite)
  "#66A61E",  # A- T+ P- (Temperature)
  "#E6AB02",  # A- T+ P+ (Temperature_Parasite)
  "#A6761D",  # A+ T+ P- (Antibiotics_Temperature)
  "#666666"   # A+ T+ P+ (Antibiotics_Temperature_Parasite)
)

# Create named vector for color scale
treatment_color_scale <- setNames(treatment_colors, treatment_order)

```

### Functions

```{r}

# Function to extract sample data as dataframe from phyloseq object
samdatAsDataframe <- function(ps) {
  samdat <- phyloseq::sample_data(ps)
  df <- data.frame(samdat, check.names = FALSE, stringsAsFactors = FALSE)
  return(df)
}

# Function to rename variables in phyloseq object
ps_rename <- function(ps, ...) {
  ps <- microViz::ps_get(ps)
  df <- samdatAsDataframe(ps)
  df <- dplyr::rename(.data = df, ...)
  phyloseq::sample_data(ps) <- df
  return(ps)
}

# SourceFolder function
source(here::here("Code", "R", "Functions", "StartFunctions", "sourceFolder.R"))

# Import all helper functions found in `/Functions`
sourceFolder(here::here("Code", "R", "Functions", "StartFunctions"), T)
sourceFolder(here::here("Code", "R", "Functions", "HelperFunctions"), T)
# sourceFolder(here::here("Code", "R", "Functions", "AnalysisScripts"), T)

# Source the calculate_dynamic_metrics.R script to load required functions
source(here::here("Code", "R", "Functions", "AnalysisScripts", "calculate_dynamic_metrics.R"))

```

### Import Data

```{r}

ps.tmp <- readRDS("/Users/michaelsieler/Dropbox/Mac (2)/Documents/Sharpton_Lab/Projects_Repository/Rules_of_Life/major-experiment-2023/Data/Robjects/pseq_uncleaned_05052025.rds")


ps.cleaned <-
    ps.tmp %>%
        ## Update Metadata
        ps_rename(Time = Timepoint) %>%
        microViz::ps_mutate(
            Treatment = case_when(
                Antibiotics == 0 & Temperature == 0 & Pathogen == 0 ~ "A- T- P-",
                Antibiotics == 0 & Temperature == 0 & Pathogen == 1 ~ "A- T- P+",
                Antibiotics == 1 & Temperature == 0 & Pathogen == 0 ~ "A+ T- P-",
                Antibiotics == 1 & Temperature == 0 & Pathogen == 1 ~ "A+ T- P+",
                Antibiotics == 0 & Temperature == 1 & Pathogen == 0 ~ "A- T+ P-",
                Antibiotics == 0 & Temperature == 1 & Pathogen == 1 ~ "A- T+ P+",
                Antibiotics == 1 & Temperature == 1 & Pathogen == 0 ~ "A+ T+ P-",
                Antibiotics == 1 & Temperature == 1 & Pathogen == 1 ~ "A+ T+ P+",
                TRUE ~ "Unknown"
            ), .after = "Pathogen"
        ) %>%
        microViz::ps_mutate(Sample = fecal.sample.number, .before = 1) %>%
        microViz::ps_mutate(Sample = gsub("^f", "", Sample)) %>%
        microViz::ps_filter(Treatment != "Unknown") %>%
        microViz::ps_mutate(
            History = case_when(
                Antibiotics + Temperature == 0 ~ 0,
                Antibiotics + Temperature == 1 ~ 1,
                Antibiotics + Temperature == 2 ~ 2,
            ), .after = "Treatment"
        ) %>%
        
        ## Additional metadata updates, factorizing metadata
        microViz::ps_mutate(
        # Create treatment code
            treatment_code = case_when(
              Antibiotics == 0 & Temperature == 0 & Pathogen == 0 ~ "Aneg_Tneg_Pneg",
              Antibiotics == 0 & Temperature == 0 & Pathogen == 1 ~ "Aneg_Tneg_Ppos",
              Antibiotics == 1 & Temperature == 0 & Pathogen == 0 ~ "Apos_Tneg_Pneg",
              Antibiotics == 1 & Temperature == 0 & Pathogen == 1 ~ "Apos_Tneg_Ppos",
              Antibiotics == 0 & Temperature == 1 & Pathogen == 0 ~ "Aneg_Tpos_Pneg",
              Antibiotics == 0 & Temperature == 1 & Pathogen == 1 ~ "Aneg_Tpos_Ppos",
              Antibiotics == 1 & Temperature == 1 & Pathogen == 0 ~ "Apos_Tpos_Pneg",
              Antibiotics == 1 & Temperature == 1 & Pathogen == 1 ~ "Apos_Tpos_Ppos"
            ),
            # Create treatment group factor
            treatment_group = case_when(
              Antibiotics == 0 & Temperature == 0 & Pathogen == 1 ~ "Parasite",
              Antibiotics == 1 & Temperature == 0 & Pathogen == 0 ~ "Antibiotics",
              Antibiotics == 1 & Temperature == 0 & Pathogen == 1 ~ "Antibiotics_Parasite",
              Antibiotics == 0 & Temperature == 1 & Pathogen == 0 ~ "Temperature",
              Antibiotics == 0 & Temperature == 1 & Pathogen == 1 ~ "Temperature_Parasite",
              Antibiotics == 1 & Temperature == 1 & Pathogen == 0 ~ "Antibiotics_Temperature",
              Antibiotics == 1 & Temperature == 1 & Pathogen == 1 ~ "Antibiotics_Temperature_Parasite",
              TRUE ~ "Control"
            ),
            # Convert to factor with appropriate levels
            treatment_group = factor(treatment_group, 
                                   levels = c("Control", "Parasite", 
                                              "Antibiotics", "Antibiotics_Parasite",
                                              "Temperature", "Temperature_Parasite",
                                            "Antibiotics_Temperature", "Antibiotics_Temperature_Parasite")
                                   ),
            treatment_code = factor(treatment_code, levels = treatment_order),
            # Create time point factor
            time_point = factor(Time, levels = c(0, 14, 18, 25, 29, 60)),
            # Create pathogen status factor
            pathogen_status = factor(ifelse(Pathogen == 1, "Exposed", "Unexposed"),
                                   levels = c("Unexposed", "Exposed")),
            # Create sex factor
            sex = factor(Sex, levels = c("M", "F"))
            )  %>%
  # Fix names for taxonomic ranks not identified
  microViz::tax_fix(suffix_rank = "current", anon_unique = T, unknown = NA) %>% 
  # Filter for any samples that contain more than 5000 reads
  microViz::ps_filter(sample_sums(.) > 5000) %>%
  # Any taxa not found in at least 3 samples are removed
  microViz::tax_filter(min_prevalence = 3, undetected = 0) %>%
  # Remove any unwanted reads
  microViz::tax_select(c("Mitochondria", "Chloroplast", "Eukaryota"), deselect = TRUE) %>%
  microViz::tax_select(c("Bacteria, Phylum"), deselect = TRUE) 


# ps.cleaned %>% microViz::samdat_tbl()

```

### Variables

```{r}
## Diversity Methods

diversity.method <- list()
diversity.method[["alpha"]] <- c("shannon", 
                                 "inverse_simpson", 
                                 "observed")
diversity.method[["beta"]] <- c("bray", "canberra")

## Stats/Plotting Variables

alpha.stats <- list() # save all stat results here
alpha.plots <- list() # save plots here

beta.stats <- list() # save all stat results here
beta.plots <- list() # save plots here

diffAbnd.stats <- list() # save all stat results here
diffAbnd.plots <- list() # save plots here
```


### Calculate Scores

<!-- ```{r} -->
<!-- ps.cleaned %>% # phyloseq object we'll be using -->
<!--   microViz::ps_calc_diversity( # calculates various alpha diversity metrics -->
<!--     rank = "Genus", # What taxonomic rank do you want to calculate diversity metrics at -->
<!--     index = "gini_simpson", # What diversity metric to use -->
<!--     varname = "Simpson__Genus", # Column name in sample data -->
<!--     exp = F # exponentiate the result or not -->
<!--   ) %>% -->
<!--     samdat_tbl() %>% -->
<!--     dplyr::relocate(Simpson__Genus, .after = 2) %>% arrange(desc(Simpson__Genus)) -->
<!-- ``` -->


```{r}

## Alpha
ps.cleaned_alpha <-
    ps.cleaned %>% # phyloseq object we'll be using
  microViz::ps_calc_diversity( # calculates various alpha diversity metrics
    rank = "Genus", # What taxonomic rank do you want to calculate diversity metrics at
    index = "shannon", # What diversity metric to use
    varname = "Shannon__Genus", # Column name in sample data
    exp = F # exponentiate the result or not
  ) %>%
  microViz::ps_calc_diversity(
    rank = "Genus",
    index = "gini_simpson",
    varname = "Simpson__Genus" # Column name in sample data
  ) %>%
  microViz::ps_calc_richness( # related to ps_calc_diversity, this calculates richness or "observed" values
    rank = "Genus",
    varname = "Richness__Genus"
  ) %>%
  # ps_calc_diversity.phy(
  #   varname = "Phylogenetic__Genus"
  # ) %>% # Note: This is a helper function I made. You need to adjust function manually to change taxon rank (See: microViz_Helper.R)
  # # ps_calc_diversity.phy() helper function does not properly name the column. Modify the following function as needed 
  # ps_rename("Phylogenetic__Genus" = "phylogenetic_Genus") %>% # Helper function (See: microViz_Helper.R)
  ps_mutate(across(contains("__Genus"), norm_scores, .names = "{.col}_norm")) 
    


test.data.alpha <-
    ps.cleaned_alpha %>% 
    microViz::samdat_tbl() %>%
    dplyr::relocate(c(".sample_name"), .before = "Antibiotics") %>%
    tidyr::pivot_longer(cols = contains("__Genus"), names_to = "Alpha.Metric", values_to = "Alpha.Score") %>%
    dplyr::mutate(Data.Type = case_when(
        Alpha.Metric %in% c("Simpson__Genus_norm", "Simpson__Genus_norm", "Richness__Genus_norm") ~ "Normalized",
        TRUE ~ "Raw"
    )) %>%
    dplyr::relocate(c("Alpha.Metric", "Alpha.Score", "Data.Type"), .after = "Tank.ID") %>%
    dplyr::relocate(c("Sample", "Treatment", "Time", "Tank.ID"), .before = 1)
    # tidyr::pivot_longer(cols = contains("_norm"), names_to = "Alpha.Metric", values_to = "Alpha.Score") %>%
    # dplyr::filter(Alpha.Metric %in% c("Simpson__Genus_norm", "Simpson__Genus_norm")) %>%
    # tidyr::pivot_longer(cols = contains("__Genus"), names_to = "Raw.Metric", values_to = "Raw.Score") 


```

```{r}
beta.dist.mat <- list()

beta.dist.mat[["All"]] <- # saves the results of this loop to a list
  furrr::future_map(diversity.method[["beta"]], function(x){ # Future_map runs the function in parallel
    ps.cleaned_alpha %>% # is the phyloseq object we are looping over
      microViz::tax_transform(trans = "identity", # tax transform transforms taxa counts
                    rank = ifelse(x == "gunifrac" || x == "wunifrac" || x == "unifrac", "unique", "Genus")) %>% # phylogenetic methods need "unique" ASV counts
      microViz::dist_calc(x, gunifrac_alpha = 0.5) # calculates the distance between samples. gunifrac_alpha = 0.5 weights abundance into its calculation
  }) %>%
  setNames(diversity.method[["beta"]]) # assigns names of the beta methods to the list

data.beta.list <- list()

# Save dataframe
data.beta.list[["All"]] <-
  
  ps.cleaned_alpha %>% 
  
  # Calculate Distance matrix
  microViz::dist_calc("bray") %>% 
  
  # Get distance matric
  microViz::dist_get() %>%
  # Pat Schloss methods to view beta diversity matrix data
  # Source: https://youtu.be/Gm-kg3TuML4?si=nYLLAxm7uv8aYG27&t=250
  
  # Convert into a matrix object
  as.matrix() %>%
  
  # Convert matrix into a tibble
  tibble::as_tibble(rownames = "Sample_a") %>%
  
  # Convert dataframe into "long" format
  tidyr::pivot_longer(-Sample_a, names_to = "Sample_b", values_to = "Dist") %>%
  
  # Remove any rows where names equal eachother
  dplyr::filter(Sample_a != Sample_b) %>%
  
  # Merge sample dataframe
  dplyr::right_join((ps.cleaned_alpha %>% 
                       microViz::samdat_tbl() %>% ungroup() %>%  arrange(Time, Treatment, Tank.ID) %>% 
                       dplyr::select(!contains("_Genus")) %>%
                       dplyr::select(fecal.sample.number, Time, Treatment, Tank.ID)
  ), 
  by = c("Sample_a" = "fecal.sample.number")) %>% 
  dplyr::right_join((ps.cleaned_alpha %>% 
                       microViz::samdat_tbl() %>% 
                       dplyr::ungroup() %>% 
                       dplyr::arrange(Time, Treatment, Tank.ID) %>% 
                       dplyr::select(!contains("_Genus")) %>%
                       dplyr::select(fecal.sample.number, Time, Treatment, Tank.ID)
  ), 
  by = c("Sample_b" = "fecal.sample.number"), suffix = c("_a", "_b")) %>% # add suffix to differentiate between sample metadata
  # Move columns around
  dplyr::select(order(colnames(.))) %>%
  dplyr::relocate(dplyr::any_of(c("Sample_a", "Sample_b", "Dist")), .before = 1) %>%
  dplyr::relocate(dplyr::contains(c("Tank")), .after = dplyr::last_col()) %>%
  dplyr::ungroup() %>%
  
  # Calculate the median distance of samples to all other samples (Measure of "personalization")
  dplyr::group_by(Sample_a, Treatment_a, Time_a, Tank.ID_a) %>%
  dplyr::summarize(Beta.Score = median(Dist), #*100,  
                   Beta.Score.Mean = mean(Dist), #*100, 
                   count = dplyr::n()) %>% 
  dplyr::ungroup() %>%
  
  # Rename Columns
  dplyr::rename(fecal.sample.number = Sample_a,
                Treatment = Treatment_a,
                
                Time = Time_a,
                Tank.ID = Tank.ID_a) %>%
  dplyr::right_join(ps.cleaned_alpha %>%
                      microViz::samdat_tbl() %>% 
                      dplyr::select(!contains("_Genus")),
                    by = c("fecal.sample.number", "Time", "Treatment","Tank.ID")) %>%
  dplyr::ungroup() %>%
  
  # Normalize diversity scores
  dplyr::mutate(dplyr::across(dplyr::contains("Beta.Score"), norm_scores, .names = "{.col}_norm"), .after = "Beta.Score") %>%
  
  # Create a column to save what type of beta metric was calculated
  dplyr::mutate(Beta.Metric = "Bray-Curtis", .before = "Beta.Score") %>%
  dplyr::mutate(Sample = gsub("^f", "", fecal.sample.number)) %>%
    dplyr::relocate(Sample, .before = 1) %>%
    dplyr::relocate(fecal.sample.number, .after = 36)


# Prepare beta diversity data
test.data.beta <-
    data.beta.list[["All"]] %>%
        tidyr::pivot_longer(cols = c("Beta.Score", "Beta.Score_norm"), names_to = "Beta.Metric.Type", values_to = "Beta.Score") %>%
        dplyr::relocate(c("Beta.Metric.Type", "Beta.Score"), .after = "Beta.Metric") %>%
        dplyr::mutate(Data.Type = case_when(
            Beta.Metric.Type %in% c("	Beta.Score", "Beta.Score_norm") ~ "Normalized",
            TRUE ~ "Raw"
        ), .after = "Beta.Score") %>%
        # Remove the temporary Beta.Metric.Type column
        dplyr::select(-c(Beta.Metric.Type, count)) %>%
        dplyr::mutate(Beta.Metric = ifelse(Data.Type == "Normalized", paste0(Beta.Metric, "_norm"), Beta.Metric)) %>%
        dplyr::arrange(Sample, Time, Tank.ID, Data.Type)
```




### Calculate SADMMs

```{r}

# Calculate all metrics at once using the combined function
test.alpha.all_metrics <- calculate_all_dynamic_metrics(
  data = test.data.alpha,
  metric_col = "Alpha.Metric",
  score_col = "Alpha.Score",
  ref_group = NULL,  # Using all groups as reference
  time_col = "Time",
  group_col = "Treatment",
  use_mean = TRUE
)

# Calculate all metrics at once using the combined function
test.beta.all_metrics <- calculate_all_dynamic_metrics(
  data = test.data.beta,
  metric_col = "Beta.Metric",
  score_col = "Beta.Score",
  ref_group = NULL,
  time_col = "Time",
  group_col = "Treatment",
  use_mean = TRUE
)



```

```{r}


# Merge Displacement data with Alpha.Metric and Alpha.Score
data.alpha.merged <-
    test.alpha.all_metrics$volatility %>%
        # dplyr::filter()
        # First pivot the SADMM metrics
        # dplyr::group_by(Data.Type) %>%
        tidyr::pivot_longer(
            cols = c("Alpha.Score", "Ref.Score", "Displacement", "Velocity", "Acceleration", "Volatility"),
            names_to = "METRIC",
            values_to = "SCORE"
        ) %>%
        dplyr::mutate(METRIC = factor(METRIC, levels = c("Alpha.Score", "Ref.Score", "Displacement", "Velocity", "Acceleration", "Volatility")),
                      Data.Type = factor(Data.Type, levels = c("Raw", "Normalized")))



# Merge Displacement data with Alpha.Metric and Alpha.Score
data.beta.merged <-
    test.beta.all_metrics$volatility %>%
        # dplyr::filter()
        # First pivot the SADMM metrics
        # dplyr::group_by(Data.Type) %>%
        tidyr::pivot_longer(
            cols = c("Beta.Score", "Ref.Score", "Displacement", "Velocity", "Acceleration", "Volatility"),
            names_to = "METRIC",
            values_to = "SCORE"
        ) %>%
        dplyr::mutate(METRIC = factor(METRIC, levels = c("Beta.Score", "Ref.Score", "Displacement", "Velocity", "Acceleration", "Volatility")),
                      Data.Type = factor(Data.Type, levels = c("Raw", "Normalized")))

```



## 00) All 

```{r echo=FALSE}
# Use here package to build project-relative path to the experimental design image
knitr::include_graphics(here::here("Manuscript", "Figures", "Experimental_Design", "Experimental_Design__ALL.png"))
```

### Alpha

#### {.tabset}

##### All

```{r include=FALSE}

p <- 
    data.alpha.merged %>%
        # dplyr::filter(Time == 60) %>%
        dplyr::filter(Alpha.Metric %in% c("Simpson__Genus", "Simpson__Genus_norm")) %>%
        dplyr::filter(!METRIC %in% c("Ref.Score", "Velocity", "Acceleration", "Volatility")) %>%
        # Filter out specific combination: Simpson__Genus + Raw + Alpha.Score
        dplyr::filter(!(Alpha.Metric == "Simpson__Genus" & Data.Type == "Raw" & METRIC == "Displacement")) %>%
        ggplot(aes(x = Treatment, y = SCORE, fill = Treatment)) +
        geom_hline(yintercept = 0) +
        geom_violin(draw_quantiles = c(0.5), alpha = 0.7) +
        facet_grid(METRIC+Data.Type ~ ., scales = "free") +
        scale_fill_manual(values = treatment_color_scale) +
        labs(title = "Alpha Diversity - Violin Plots") +
        theme(axis.text.x = element_text(angle = 45, hjust = 1))
```


```{r fig.height=6, fig.width=8, echo=FALSE, message=FALSE, warning=FALSE}

p

```

##### 0 DPE

```{r include=FALSE}

p <- 
    data.alpha.merged %>%
        dplyr::filter(Time == 0) %>%
        dplyr::filter(Alpha.Metric %in% c("Simpson__Genus", "Simpson__Genus_norm")) %>%
        dplyr::filter(!METRIC %in% c("Ref.Score", "Velocity", "Acceleration", "Volatility")) %>%
        # Filter out specific combination: Simpson__Genus + Raw + Alpha.Score
        dplyr::filter(!(Alpha.Metric == "Simpson__Genus" & Data.Type == "Raw" & METRIC == "Displacement")) %>%
        ggplot(aes(x = Treatment, y = SCORE, fill = Treatment)) +
        geom_hline(yintercept = 0) +
        geom_violin(draw_quantiles = c(0.5), alpha = 0.7) +
        facet_grid(METRIC+Data.Type ~ ., scales = "free") +
        scale_fill_manual(values = treatment_color_scale) +
        labs(title = "Alpha Diversity - Violin Plots") +
        theme(axis.text.x = element_text(angle = 45, hjust = 1))
```


```{r fig.height=6, fig.width=8, echo=FALSE, message=FALSE, warning=FALSE}

p

```

##### 14 DPE

```{r include=FALSE}

p <- 
    data.alpha.merged %>%
        dplyr::filter(Time == 14) %>%
        dplyr::filter(Alpha.Metric %in% c("Simpson__Genus", "Simpson__Genus_norm")) %>%
        dplyr::filter(!METRIC %in% c("Ref.Score", "Velocity", "Acceleration", "Volatility")) %>%
        # Filter out specific combination: Simpson__Genus + Raw + Alpha.Score
        dplyr::filter(!(Alpha.Metric == "Simpson__Genus" & Data.Type == "Raw" & METRIC == "Displacement")) %>%
        ggplot(aes(x = Treatment, y = SCORE, fill = Treatment)) +
        geom_hline(yintercept = 0) +
        geom_violin(draw_quantiles = c(0.5), alpha = 0.7) +
        facet_grid(METRIC+Data.Type ~ ., scales = "free") +
        scale_fill_manual(values = treatment_color_scale) +
        labs(title = "Alpha Diversity - Violin Plots") +
        theme(axis.text.x = element_text(angle = 45, hjust = 1))
```


```{r fig.height=6, fig.width=8, echo=FALSE, message=FALSE, warning=FALSE}

p

```

##### 18 DPE

```{r include=FALSE}

p <- 
    data.alpha.merged %>%
        dplyr::filter(Time == 18) %>%
        dplyr::filter(Alpha.Metric %in% c("Simpson__Genus", "Simpson__Genus_norm")) %>%
        dplyr::filter(!METRIC %in% c("Ref.Score", "Velocity", "Acceleration", "Volatility")) %>%
        # Filter out specific combination: Simpson__Genus + Raw + Alpha.Score
        dplyr::filter(!(Alpha.Metric == "Simpson__Genus" & Data.Type == "Raw" & METRIC == "Displacement")) %>%
        ggplot(aes(x = Treatment, y = SCORE, fill = Treatment)) +
        geom_hline(yintercept = 0) +
        geom_violin(draw_quantiles = c(0.5), alpha = 0.7) +
        facet_grid(METRIC+Data.Type ~ ., scales = "free") +
        scale_fill_manual(values = treatment_color_scale) +
        labs(title = "Alpha Diversity - Violin Plots") +
        theme(axis.text.x = element_text(angle = 45, hjust = 1))
```


```{r fig.height=6, fig.width=8, echo=FALSE, message=FALSE, warning=FALSE}

p

```

##### 25 DPE

```{r include=FALSE}

p <- 
    data.alpha.merged %>%
        dplyr::filter(Time == 25) %>%
        dplyr::filter(Alpha.Metric %in% c("Simpson__Genus", "Simpson__Genus_norm")) %>%
        dplyr::filter(!METRIC %in% c("Ref.Score", "Velocity", "Acceleration", "Volatility")) %>%
        # Filter out specific combination: Simpson__Genus + Raw + Alpha.Score
        dplyr::filter(!(Alpha.Metric == "Simpson__Genus" & Data.Type == "Raw" & METRIC == "Displacement")) %>%
        ggplot(aes(x = Treatment, y = SCORE, fill = Treatment)) +
        geom_hline(yintercept = 0) +
        geom_violin(draw_quantiles = c(0.5), alpha = 0.7) +
        facet_grid(METRIC+Data.Type ~ ., scales = "free") +
        scale_fill_manual(values = treatment_color_scale) +
        labs(title = "Alpha Diversity - Violin Plots") +
        theme(axis.text.x = element_text(angle = 45, hjust = 1))
```


```{r fig.height=6, fig.width=8, echo=FALSE, message=FALSE, warning=FALSE}

p

```

##### 29 DPE

```{r include=FALSE}

p <- 
    data.alpha.merged %>%
        dplyr::filter(Time == 29) %>%
        dplyr::filter(Alpha.Metric %in% c("Simpson__Genus", "Simpson__Genus_norm")) %>%
        dplyr::filter(!METRIC %in% c("Ref.Score", "Velocity", "Acceleration", "Volatility")) %>%
        # Filter out specific combination: Simpson__Genus + Raw + Alpha.Score
        dplyr::filter(!(Alpha.Metric == "Simpson__Genus" & Data.Type == "Raw" & METRIC == "Displacement")) %>%
        ggplot(aes(x = Treatment, y = SCORE, fill = Treatment)) +
        geom_hline(yintercept = 0) +
        geom_violin(draw_quantiles = c(0.5), alpha = 0.7) +
        facet_grid(METRIC+Data.Type ~ ., scales = "free") +
        scale_fill_manual(values = treatment_color_scale) +
        labs(title = "Alpha Diversity - Violin Plots") +
        theme(axis.text.x = element_text(angle = 45, hjust = 1))
```


```{r fig.height=6, fig.width=8, echo=FALSE, message=FALSE, warning=FALSE}

p

```

##### 60 DPE

```{r include=FALSE}

p <- 
    data.alpha.merged %>%
        dplyr::filter(Time == 60) %>%
        dplyr::filter(Alpha.Metric %in% c("Simpson__Genus", "Simpson__Genus_norm")) %>%
        dplyr::filter(!METRIC %in% c("Ref.Score", "Velocity", "Acceleration", "Volatility")) %>%
        # Filter out specific combination: Simpson__Genus + Raw + Alpha.Score
        dplyr::filter(!(Alpha.Metric == "Simpson__Genus" & Data.Type == "Raw" & METRIC == "Displacement")) %>%
        ggplot(aes(x = Treatment, y = SCORE, fill = Treatment)) +
        geom_hline(yintercept = 0) +
        geom_violin(draw_quantiles = c(0.5), alpha = 0.7) +
        facet_grid(METRIC+Data.Type ~ ., scales = "free") +
        scale_fill_manual(values = treatment_color_scale) +
        labs(title = "Alpha Diversity - Violin Plots") +
        theme(axis.text.x = element_text(angle = 45, hjust = 1))
```


```{r fig.height=6, fig.width=8, echo=FALSE, message=FALSE, warning=FALSE}

p

```

###

#### By Treatment {.tabset}

```{r include=FALSE}

# Create separate plots for Raw and Normalized data with different y-axis scales
p <-
data.alpha.merged %>%
    dplyr::filter(Alpha.Metric %in% c("Simpson__Genus", "Simpson__Genus_norm")) %>%
    dplyr::filter(!METRIC %in% c("Ref.Score", "Velocity", "Acceleration", "Volatility")) %>%
    # Filter out specific combination: Simpson__Genus + Raw + Alpha.Score
    dplyr::filter(!(Alpha.Metric == "Simpson__Genus" & Data.Type == "Raw" & METRIC == "Displacement")) %>%
    # Ensure Treatment is a factor with the correct order
    dplyr::mutate(Treatment = factor(Treatment, levels = treatment_order)) %>%
    # Group by Treatment to create separate plots
    dplyr::group_by(Treatment) %>%
    dplyr::group_map(~ {
        # Get the treatment color for this group
        treatment_color <- treatment_color_scale[.y$Treatment]
        
        # Create separate plots for each Data.Type
        raw_alpha_plot <- ggplot(.x %>% dplyr::filter(Alpha.Metric == "Simpson__Genus" & Data.Type == "Raw" & METRIC == "Alpha.Score"), 
                          aes(x = Time, y = SCORE)) +
            geom_point(alpha = 0.5, color = treatment_color) +
            geom_smooth(method = "loess", se = F, size = 1, color = treatment_color) +
            # facet_wrap(.~Tank.ID) +
            scale_y_continuous(limits = c(0, 1)) +
            labs(title = paste("Treatment:", .y$Treatment, " (Raw Alpha Score)"))
        
        norm_alpha_plot <- ggplot(.x %>% dplyr::filter(Alpha.Metric == "Simpson__Genus_norm" & Data.Type == "Normalized" & METRIC == "Alpha.Score"), 
                           aes(x = Time, y = SCORE)) +
            geom_point(alpha = 0.5, color = treatment_color) +
            geom_smooth(method = "loess", se = F, size = 1, color = treatment_color) +
            # facet_wrap(.~Tank.ID) +
            scale_y_continuous(limits = c(0, 1)) +
            labs(title = paste("Treatment:", .y$Treatment, " (Normalized Alpha Score)"))
        
        # Create plot for Normalized Displacement
        norm_disp_plot <- ggplot(.x %>% dplyr::filter(Alpha.Metric == "Simpson__Genus_norm" & Data.Type == "Normalized" & METRIC == "Displacement"), 
                           aes(x = Time, y = SCORE)) +
            geom_point(alpha = 0.5, color = treatment_color) +
            geom_smooth(method = "loess", se = F, size = 1, color = treatment_color) +
            # facet_wrap(.~Tank.ID) +
            scale_y_continuous(limits = c(-.667, .667)) +
            labs(title = paste("Treatment:", .y$Treatment, " (Normalized Displacement)"))
        
        # Return all three plots as a named list
        list(raw = raw_alpha_plot, norm = norm_alpha_plot, disp = norm_disp_plot)
    }) %>% setNames(treatment_order)

```



##### (hide)

```{r}

```


##### A- T- P- (Control)

```{r fig.height=6, fig.width=4, echo=FALSE, message=FALSE, warning=FALSE}
p[["A- T- P-"]]$raw

p[["A- T- P-"]]$norm

p[["A- T- P-"]]$disp
```

##### A- T- P+ (Parasite)

```{r fig.width=4, fig.height=6, echo=FALSE, message=FALSE, warning=FALSE}
p[["A- T- P+"]]$raw
 
p[["A- T- P+"]]$norm

p[["A- T- P+"]]$disp
```

##### A+ T- P- (Antibiotics)

```{r fig.width=4, fig.height=6, echo=FALSE, message=FALSE, warning=FALSE}
p[["A+ T- P-"]]$raw

p[["A+ T- P-"]]$norm

p[["A+ T- P-"]]$disp
```

##### A+ T- P+ (Antibiotics_Parasite)

```{r fig.width=4, fig.height=6, echo=FALSE, message=FALSE, warning=FALSE}
p[["A+ T- P+"]]$raw

p[["A+ T- P+"]]$norm

p[["A+ T- P+"]]$disp
```

##### A- T+ P- (Temperature)

```{r fig.width=4, fig.height=6, echo=FALSE, message=FALSE, warning=FALSE}
p[["A- T+ P-"]]$raw

p[["A- T+ P-"]]$norm

p[["A- T+ P-"]]$disp
```

##### A- T+ P+ (Temperature_Parasite)

```{r fig.width=4, fig.height=6, echo=FALSE, message=FALSE, warning=FALSE}
p[["A- T+ P+"]]$raw

p[["A- T+ P+"]]$norm

p[["A- T+ P+"]]$disp
```

##### A+ T+ P- (Antibiotics_Temperature)

```{r fig.width=4, fig.height=6, echo=FALSE, message=FALSE, warning=FALSE}
p[["A+ T+ P-"]]$raw

p[["A+ T+ P-"]]$norm

p[["A+ T+ P-"]]$disp
```

##### A+ T+ P+ (Antibiotics_Temperature_Parasite)

```{r fig.width=4, fig.height=6, echo=FALSE, message=FALSE, warning=FALSE}
p[["A+ T+ P+"]]$raw

p[["A+ T+ P+"]]$norm

p[["A+ T+ P+"]]$disp
```


#### By Tank {.tabset}

```{r include=FALSE}

# Create separate plots for Raw and Normalized data with different y-axis scales
p <-
data.alpha.merged %>%
    dplyr::filter(Alpha.Metric %in% c("Simpson__Genus", "Simpson__Genus_norm")) %>%
    dplyr::filter(!METRIC %in% c("Ref.Score", "Velocity", "Acceleration", "Volatility")) %>%
    # Filter out specific combination: Simpson__Genus + Raw + Alpha.Score
    dplyr::filter(!(Alpha.Metric == "Simpson__Genus" & Data.Type == "Raw" & METRIC == "Displacement")) %>%
    # Ensure Treatment is a factor with the correct order
    dplyr::mutate(Treatment = factor(Treatment, levels = treatment_order)) %>%
    # Group by Treatment to create separate plots
    dplyr::group_by(Treatment) %>%
    dplyr::group_map(~ {
        # Get the treatment color for this group
        treatment_color <- treatment_color_scale[.y$Treatment]
        
        # Create separate plots for each Data.Type
        raw_alpha_plot <- ggplot(.x %>% dplyr::filter(Alpha.Metric == "Simpson__Genus" & Data.Type == "Raw" & METRIC == "Alpha.Score"), 
                          aes(x = Time, y = SCORE, group = Tank.ID)) +
            geom_point(alpha = 0.5, color = treatment_color) +
            geom_smooth(method = "loess", se = F, size = 1, color = treatment_color) +
            facet_wrap(.~Tank.ID) +
            scale_y_continuous(limits = c(0, 1)) +
            labs(title = paste("Treatment:", .y$Treatment, " (Raw Alpha Score)"))
        
        norm_alpha_plot <- ggplot(.x %>% dplyr::filter(Alpha.Metric == "Simpson__Genus_norm" & Data.Type == "Normalized" & METRIC == "Alpha.Score"), 
                           aes(x = Time, y = SCORE, group = Tank.ID)) +
            geom_point(alpha = 0.5, color = treatment_color) +
            geom_smooth(method = "loess", se = F, size = 1, color = treatment_color) +
            facet_wrap(.~Tank.ID) +
            scale_y_continuous(limits = c(0, 1)) +
            labs(title = paste("Treatment:", .y$Treatment, " (Normalized Alpha Score)"))
        
        # Create plot for Normalized Displacement
        norm_disp_plot <- ggplot(.x %>% dplyr::filter(Alpha.Metric == "Simpson__Genus_norm" & Data.Type == "Normalized" & METRIC == "Displacement"), 
                           aes(x = Time, y = SCORE, group = Tank.ID)) +
            geom_point(alpha = 0.5, color = treatment_color) +
            geom_smooth(method = "loess", se = F, size = 1, color = treatment_color) +
            facet_wrap(.~Tank.ID) +
            scale_y_continuous(limits = c(-.667, .667)) +
            labs(title = paste("Treatment:", .y$Treatment, " (Normalized Displacement)"))
        
        # Return all three plots as a named list
        list(raw = raw_alpha_plot, norm = norm_alpha_plot, disp = norm_disp_plot)
    }) %>% setNames(treatment_order)
    # setNames(treatment_order[treatment_order %in% unique(data.alpha.merged$Treatment)])

```

##### (hide)

```{r}

```


##### A- T- P- (Control)

```{r fig.height=6, fig.width=4, echo=FALSE, message=FALSE, warning=FALSE}
p[["A- T- P-"]]$raw

p[["A- T- P-"]]$norm

p[["A- T- P-"]]$disp
```

##### A- T- P+ (Parasite)

```{r fig.width=4, fig.height=6, echo=FALSE, message=FALSE, warning=FALSE}
p[["A- T- P+"]]$raw
 
p[["A- T- P+"]]$norm

p[["A- T- P+"]]$disp
```

##### A+ T- P- (Antibiotics)

```{r fig.width=4, fig.height=6, echo=FALSE, message=FALSE, warning=FALSE}
p[["A+ T- P-"]]$raw

p[["A+ T- P-"]]$norm

p[["A+ T- P-"]]$disp
```

##### A+ T- P+ (Antibiotics_Parasite)

```{r fig.width=4, fig.height=6, echo=FALSE, message=FALSE, warning=FALSE}
p[["A+ T- P+"]]$raw

p[["A+ T- P+"]]$norm

p[["A+ T- P+"]]$disp
```

##### A- T+ P- (Temperature)

```{r fig.width=4, fig.height=6, echo=FALSE, message=FALSE, warning=FALSE}
p[["A- T+ P-"]]$raw

p[["A- T+ P-"]]$norm

p[["A- T+ P-"]]$disp
```

##### A- T+ P+ (Temperature_Parasite)

```{r fig.width=4, fig.height=6, echo=FALSE, message=FALSE, warning=FALSE}
p[["A- T+ P+"]]$raw

p[["A- T+ P+"]]$norm

p[["A- T+ P+"]]$disp
```

##### A+ T+ P- (Antibiotics_Temperature)

```{r fig.width=4, fig.height=6, echo=FALSE, message=FALSE, warning=FALSE}
p[["A+ T+ P-"]]$raw

p[["A+ T+ P-"]]$norm

p[["A+ T+ P-"]]$disp
```

##### A+ T+ P+ (Antibiotics_Temperature_Parasite)

```{r fig.width=4, fig.height=6, echo=FALSE, message=FALSE, warning=FALSE}
p[["A+ T+ P+"]]$raw

p[["A+ T+ P+"]]$norm

p[["A+ T+ P+"]]$disp
```

### Beta

#### {.tabset}

##### All

```{r include=FALSE}

p <- 
    data.beta.merged %>%
        # dplyr::filter(Time == 60) %>%
        dplyr::filter(Beta.Metric %in% c("Bray-Curtis", "Bray-Curtis_norm")) %>%
        dplyr::filter(!METRIC %in% c("Ref.Score", "Velocity", "Acceleration", "Volatility")) %>%
        dplyr::filter(!(Beta.Metric == "Bray-Curtis" & Data.Type == "Raw" & METRIC == "Displacement")) %>%
    
        ggplot(aes(x = Treatment, y = SCORE, fill = Treatment)) +
        geom_hline(yintercept = 0) +
        geom_violin(draw_quantiles = c(0.5), alpha = 0.7) +
        facet_grid(METRIC+Data.Type ~ ., scales = "free") +
        scale_fill_manual(values = treatment_color_scale) +
        labs(title = "Beta Diversity - Violin Plots") +
        theme(axis.text.x = element_text(angle = 45, hjust = 1))
```


```{r fig.height=6, fig.width=8, echo=FALSE, message=FALSE, warning=FALSE}

p

```

##### 0 DPE

```{r include=FALSE}

p <- 
    data.beta.merged %>%
        dplyr::filter(Time == 0) %>%
        dplyr::filter(Beta.Metric %in% c("Bray-Curtis", "Bray-Curtis_norm")) %>%
        dplyr::filter(!METRIC %in% c("Ref.Score", "Velocity", "Acceleration", "Volatility")) %>%
        dplyr::filter(!(Beta.Metric == "Bray-Curtis" & Data.Type == "Raw" & METRIC == "Displacement")) %>%
    
        ggplot(aes(x = Treatment, y = SCORE, fill = Treatment)) +
        geom_hline(yintercept = 0) +
        geom_violin(draw_quantiles = c(0.5), alpha = 0.7) +
        facet_grid(METRIC+Data.Type ~ ., scales = "free") +
        scale_fill_manual(values = treatment_color_scale) +
        labs(title = "Beta Diversity - Violin Plots") +
        theme(axis.text.x = element_text(angle = 45, hjust = 1))
```


```{r fig.height=6, fig.width=8, echo=FALSE, message=FALSE, warning=FALSE}

p

```

##### 14 DPE

```{r include=FALSE}

p <- 
    data.beta.merged %>%
        dplyr::filter(Time == 14) %>%
        dplyr::filter(Beta.Metric %in% c("Bray-Curtis", "Bray-Curtis_norm")) %>%
        dplyr::filter(!METRIC %in% c("Ref.Score", "Velocity", "Acceleration", "Volatility")) %>%
        dplyr::filter(!(Beta.Metric == "Bray-Curtis" & Data.Type == "Raw" & METRIC == "Displacement")) %>%
    
        ggplot(aes(x = Treatment, y = SCORE, fill = Treatment)) +
        geom_hline(yintercept = 0) +
        geom_violin(draw_quantiles = c(0.5), alpha = 0.7) +
        facet_grid(METRIC+Data.Type ~ ., scales = "free") +
        scale_fill_manual(values = treatment_color_scale) +
        labs(title = "Beta Diversity - Violin Plots") +
        theme(axis.text.x = element_text(angle = 45, hjust = 1))
```


```{r fig.height=6, fig.width=8, echo=FALSE, message=FALSE, warning=FALSE}

p

```

##### 18 DPE

```{r include=FALSE}

p <- 
    data.beta.merged %>%
        dplyr::filter(Time == 18) %>%
        dplyr::filter(Beta.Metric %in% c("Bray-Curtis", "Bray-Curtis_norm")) %>%
        dplyr::filter(!METRIC %in% c("Ref.Score", "Velocity", "Acceleration", "Volatility")) %>%
        dplyr::filter(!(Beta.Metric == "Bray-Curtis" & Data.Type == "Raw" & METRIC == "Displacement")) %>%
    
        ggplot(aes(x = Treatment, y = SCORE, fill = Treatment)) +
        geom_hline(yintercept = 0) +
        geom_violin(draw_quantiles = c(0.5), alpha = 0.7) +
        facet_grid(METRIC+Data.Type ~ ., scales = "free") +
        scale_fill_manual(values = treatment_color_scale) +
        labs(title = "Beta Diversity - Violin Plots") +
        theme(axis.text.x = element_text(angle = 45, hjust = 1))
```


```{r fig.height=6, fig.width=8, echo=FALSE, message=FALSE, warning=FALSE}

p

```

##### 25 DPE

```{r include=FALSE}

p <- 
    data.beta.merged %>%
        dplyr::filter(Time == 25) %>%
        dplyr::filter(Beta.Metric %in% c("Bray-Curtis", "Bray-Curtis_norm")) %>%
        dplyr::filter(!METRIC %in% c("Ref.Score", "Velocity", "Acceleration", "Volatility")) %>%
        dplyr::filter(!(Beta.Metric == "Bray-Curtis" & Data.Type == "Raw" & METRIC == "Displacement")) %>%
    
        ggplot(aes(x = Treatment, y = SCORE, fill = Treatment)) +
        geom_hline(yintercept = 0) +
        geom_violin(draw_quantiles = c(0.5), alpha = 0.7) +
        facet_grid(METRIC+Data.Type ~ ., scales = "free") +
        scale_fill_manual(values = treatment_color_scale) +
        labs(title = "Beta Diversity - Violin Plots") +
        theme(axis.text.x = element_text(angle = 45, hjust = 1))
```


```{r fig.height=6, fig.width=8, echo=FALSE, message=FALSE, warning=FALSE}

p

```

##### 29 DPE

```{r include=FALSE}

p <- 
    data.beta.merged %>%
        dplyr::filter(Time == 29) %>%
        dplyr::filter(Beta.Metric %in% c("Bray-Curtis", "Bray-Curtis_norm")) %>%
        dplyr::filter(!METRIC %in% c("Ref.Score", "Velocity", "Acceleration", "Volatility")) %>%
        dplyr::filter(!(Beta.Metric == "Bray-Curtis" & Data.Type == "Raw" & METRIC == "Displacement")) %>%
    
        ggplot(aes(x = Treatment, y = SCORE, fill = Treatment)) +
        geom_hline(yintercept = 0) +
        geom_violin(draw_quantiles = c(0.5), alpha = 0.7) +
        facet_grid(METRIC+Data.Type ~ ., scales = "free") +
        scale_fill_manual(values = treatment_color_scale) +
        labs(title = "Beta Diversity - Violin Plots") +
        theme(axis.text.x = element_text(angle = 45, hjust = 1))
```


```{r fig.height=6, fig.width=8, echo=FALSE, message=FALSE, warning=FALSE}

p

```

##### 60 DPE

```{r include=FALSE}

p <- 
    data.beta.merged %>%
        dplyr::filter(Time == 60) %>%
        dplyr::filter(Beta.Metric %in% c("Bray-Curtis", "Bray-Curtis_norm")) %>%
        dplyr::filter(!METRIC %in% c("Ref.Score", "Velocity", "Acceleration", "Volatility")) %>%
        dplyr::filter(!(Beta.Metric == "Bray-Curtis" & Data.Type == "Raw" & METRIC == "Displacement")) %>%
    
        ggplot(aes(x = Treatment, y = SCORE, fill = Treatment)) +
        geom_hline(yintercept = 0) +
        geom_violin(draw_quantiles = c(0.5), alpha = 0.7) +
        facet_grid(METRIC+Data.Type ~ ., scales = "free") +
        scale_fill_manual(values = treatment_color_scale) +
        labs(title = "Beta Diversity - Violin Plots") +
        theme(axis.text.x = element_text(angle = 45, hjust = 1))
```


```{r fig.height=6, fig.width=8, echo=FALSE, message=FALSE, warning=FALSE}

p

```

###

#### By Treatment {.tabset}

```{r include=FALSE}

# Create separate plots for Raw and Normalized data with different y-axis scales
p <-
data.beta.merged %>%
    dplyr::filter(Beta.Metric %in% c("Bray-Curtis", "Bray-Curtis_norm")) %>%
    dplyr::filter(!METRIC %in% c("Ref.Score", "Velocity", "Acceleration", "Volatility")) %>%
    # Filter out specific combination: Simpson__Genus + Raw + Alpha.Score
    dplyr::filter(!(Beta.Metric == "Bray-Curtis" & Data.Type == "Raw" & METRIC == "Displacement")) %>%
    # Ensure Treatment is a factor with the correct order
    dplyr::mutate(Treatment = factor(Treatment, levels = treatment_order)) %>%
    # Group by Treatment to create separate plots
    dplyr::group_by(Treatment) %>%
    dplyr::group_map(~ {
        # Get the treatment color for this group
        treatment_color <- treatment_color_scale[.y$Treatment]
        
        # Create separate plots for each Data.Type
        raw_beta_plot <- ggplot(.x %>% dplyr::filter(Beta.Metric == "Bray-Curtis" & Data.Type == "Raw" & METRIC == "Beta.Score"), 
                          aes(x = Time, y = SCORE)) +
            geom_point(alpha = 0.5, color = treatment_color) +
            geom_smooth(method = "loess", se = F, size = 1, color = treatment_color) +
            # facet_wrap(.~Tank.ID) +
            scale_y_continuous(limits = c(0, 1)) +
            labs(title = paste("Treatment:", .y$Treatment, " (Raw Beta Score)"))
        
        norm_beta_plot <- ggplot(.x %>% dplyr::filter(Beta.Metric == "Bray-Curtis_norm" & Data.Type == "Normalized" & METRIC == "Beta.Score"), 
                           aes(x = Time, y = SCORE)) +
            geom_point(alpha = 0.5, color = treatment_color) +
            geom_smooth(method = "loess", se = F, size = 1, color = treatment_color) +
            # facet_wrap(.~Tank.ID) +
            scale_y_continuous(limits = c(0, 1)) +
            labs(title = paste("Treatment:", .y$Treatment, " (Normalized Alpha Score)"))
        
        # Create plot for Normalized Displacement
        norm_disp_plot <- ggplot(.x %>% dplyr::filter(Beta.Metric == "Bray-Curtis_norm" & Data.Type == "Normalized" & METRIC == "Displacement"), 
                           aes(x = Time, y = SCORE)) +
            geom_point(alpha = 0.5, color = treatment_color) +
            geom_smooth(method = "loess", se = F, size = 1, color = treatment_color) +
            # facet_wrap(.~Tank.ID) +
            scale_y_continuous(limits = c(-.667, .667)) +
            labs(title = paste("Treatment:", .y$Treatment, " (Normalized Displacement)"))
        
        # Return all three plots as a named list
        list(raw = raw_beta_plot, norm = norm_beta_plot, disp = norm_disp_plot)
    }) %>% setNames(treatment_order)

```



##### (hide)

```{r}

```


##### A- T- P- (Control)

```{r fig.height=6, fig.width=4, echo=FALSE, message=FALSE, warning=FALSE}
p[["A- T- P-"]]$raw

p[["A- T- P-"]]$norm

p[["A- T- P-"]]$disp
```

##### A- T- P+ (Parasite)

```{r fig.width=4, fig.height=6, echo=FALSE, message=FALSE, warning=FALSE}
p[["A- T- P+"]]$raw
 
p[["A- T- P+"]]$norm

p[["A- T- P+"]]$disp
```

##### A+ T- P- (Antibiotics)

```{r fig.width=4, fig.height=6, echo=FALSE, message=FALSE, warning=FALSE}
p[["A+ T- P-"]]$raw

p[["A+ T- P-"]]$norm

p[["A+ T- P-"]]$disp
```

##### A+ T- P+ (Antibiotics_Parasite)

```{r fig.width=4, fig.height=6, echo=FALSE, message=FALSE, warning=FALSE}
p[["A+ T- P+"]]$raw

p[["A+ T- P+"]]$norm

p[["A+ T- P+"]]$disp
```

##### A- T+ P- (Temperature)

```{r fig.width=4, fig.height=6, echo=FALSE, message=FALSE, warning=FALSE}
p[["A- T+ P-"]]$raw

p[["A- T+ P-"]]$norm

p[["A- T+ P-"]]$disp
```

##### A- T+ P+ (Temperature_Parasite)

```{r fig.width=4, fig.height=6, echo=FALSE, message=FALSE, warning=FALSE}
p[["A- T+ P+"]]$raw

p[["A- T+ P+"]]$norm

p[["A- T+ P+"]]$disp
```

##### A+ T+ P- (Antibiotics_Temperature)

```{r fig.width=4, fig.height=6, echo=FALSE, message=FALSE, warning=FALSE}
p[["A+ T+ P-"]]$raw

p[["A+ T+ P-"]]$norm

p[["A+ T+ P-"]]$disp
```

##### A+ T+ P+ (Antibiotics_Temperature_Parasite)

```{r fig.width=4, fig.height=6, echo=FALSE, message=FALSE, warning=FALSE}
p[["A+ T+ P+"]]$raw

p[["A+ T+ P+"]]$norm

p[["A+ T+ P+"]]$disp
```


#### By Tank {.tabset}

```{r include=FALSE}

# Create separate plots for Raw and Normalized data with different y-axis scales
p <-
data.beta.merged %>%
    dplyr::filter(Beta.Metric %in% c("Bray-Curtis", "Bray-Curtis_norm")) %>%
    dplyr::filter(!METRIC %in% c("Ref.Score", "Velocity", "Acceleration", "Volatility")) %>%
    # Filter out specific combination: Bray-Curtis + Raw + Alpha.Score
    dplyr::filter(!(Beta.Metric == "Bray-Curtis" & Data.Type == "Raw" & METRIC == "Displacement")) %>%
    # Ensure Treatment is a factor with the correct order
    dplyr::mutate(Treatment = factor(Treatment, levels = treatment_order)) %>%
    # Group by Treatment to create separate plots
    dplyr::group_by(Treatment) %>%
    dplyr::group_map(~ {
        # Get the treatment color for this group
        treatment_color <- treatment_color_scale[.y$Treatment]
        
        # Create separate plots for each Data.Type
        raw_beta_plot <- ggplot(.x %>% dplyr::filter(Beta.Metric == "Bray-Curtis" & Data.Type == "Raw" & METRIC == "Beta.Score"), 
                          aes(x = Time, y = SCORE, group = Tank.ID)) +
            geom_point(alpha = 0.5, color = treatment_color) +
            geom_smooth(method = "loess", se = F, size = 1, color = treatment_color) +
            facet_wrap(.~Tank.ID) +
            scale_y_continuous(limits = c(0, 1)) +
            labs(title = paste("Treatment:", .y$Treatment, " (Raw Alpha Score)"))
        
        norm_beta_plot <- ggplot(.x %>% dplyr::filter(Beta.Metric == "Bray-Curtis_norm" & Data.Type == "Normalized" & METRIC == "Beta.Score"), 
                           aes(x = Time, y = SCORE, group = Tank.ID)) +
            geom_point(alpha = 0.5, color = treatment_color) +
            geom_smooth(method = "loess", se = F, size = 1, color = treatment_color) +
            facet_wrap(.~Tank.ID) +
            scale_y_continuous(limits = c(0, 1)) +
            labs(title = paste("Treatment:", .y$Treatment, " (Normalized Alpha Score)"))
        
        # Create plot for Normalized Displacement
        norm_disp_plot <- ggplot(.x %>% dplyr::filter(Beta.Metric == "Bray-Curtis_norm" & Data.Type == "Normalized" & METRIC == "Displacement"), 
                           aes(x = Time, y = SCORE, group = Tank.ID)) +
            geom_point(alpha = 0.5, color = treatment_color) +
            geom_smooth(method = "loess", se = F, size = 1, color = treatment_color) +
            facet_wrap(.~Tank.ID) +
            scale_y_continuous(limits = c(-.667, .667)) +
            labs(title = paste("Treatment:", .y$Treatment, " (Normalized Displacement)"))
        
        # Return all three plots as a named list
        list(raw = raw_beta_plot, norm = norm_beta_plot, disp = norm_disp_plot)
    }) %>% setNames(treatment_order)
    # setNames(treatment_order[treatment_order %in% unique(data.alpha.merged$Treatment)])

```

##### (hide)

```{r}

```


##### A- T- P- (Control)

```{r fig.height=6, fig.width=4, echo=FALSE, message=FALSE, warning=FALSE}
p[["A- T- P-"]]$raw

p[["A- T- P-"]]$norm

p[["A- T- P-"]]$disp
```

##### A- T- P+ (Parasite)

```{r fig.width=4, fig.height=6, echo=FALSE, message=FALSE, warning=FALSE}
p[["A- T- P+"]]$raw
 
p[["A- T- P+"]]$norm

p[["A- T- P+"]]$disp
```

##### A+ T- P- (Antibiotics)

```{r fig.width=4, fig.height=6, echo=FALSE, message=FALSE, warning=FALSE}
p[["A+ T- P-"]]$raw

p[["A+ T- P-"]]$norm

p[["A+ T- P-"]]$disp
```

##### A+ T- P+ (Antibiotics_Parasite)

```{r fig.width=4, fig.height=6, echo=FALSE, message=FALSE, warning=FALSE}
p[["A+ T- P+"]]$raw

p[["A+ T- P+"]]$norm

p[["A+ T- P+"]]$disp
```

##### A- T+ P- (Temperature)

```{r fig.width=4, fig.height=6, echo=FALSE, message=FALSE, warning=FALSE}
p[["A- T+ P-"]]$raw

p[["A- T+ P-"]]$norm

p[["A- T+ P-"]]$disp
```

##### A- T+ P+ (Temperature_Parasite)

```{r fig.width=4, fig.height=6, echo=FALSE, message=FALSE, warning=FALSE}
p[["A- T+ P+"]]$raw

p[["A- T+ P+"]]$norm

p[["A- T+ P+"]]$disp
```

##### A+ T+ P- (Antibiotics_Temperature)

```{r fig.width=4, fig.height=6, echo=FALSE, message=FALSE, warning=FALSE}
p[["A+ T+ P-"]]$raw

p[["A+ T+ P-"]]$norm

p[["A+ T+ P-"]]$disp
```

##### A+ T+ P+ (Antibiotics_Temperature_Parasite)

```{r fig.width=4, fig.height=6, echo=FALSE, message=FALSE, warning=FALSE}
p[["A+ T+ P+"]]$raw

p[["A+ T+ P+"]]$norm

p[["A+ T+ P+"]]$disp
```




